pkg/scheduler/objects/allocation_ask.go

```diff
@@ -188,6 +197,48 @@ func (aa *AllocationAsk) GetAllocatedResource() *resources.Resource {
 	return aa.AllocatedResource
 }
 
+// LogSchedulingFailure keeps track of preconditions not being met for an allocation
+func (aa *AllocationAsk) LogAllocationFailure(message string, allocate bool) {
+	// for now, don't log reservations
+	if !allocate {
+		return
+	}
+
+	aa.Lock()
+	defer aa.Unlock()
+
+	entry, ok := aa.allocLog[message]
+	if !ok {
+		entry = &AllocationLogEntry{
+			Message: message,
+		}
+		aa.allocLog[message] = entry
+	}
+	entry.Timestamp = time.Now()
+	entry.Count++
+}
+
+// GetSchedulingLog returns a list of log entries corresponding to allocation preconditions not being met
+func (aa *AllocationAsk) GetAllocationLog() []*AllocationLogEntry {
+	aa.RLock()
+	defer aa.RUnlock()
+
+	res := make([]*AllocationLogEntry, len(aa.allocLog))
+	i := 0
+	for _, log := range aa.allocLog {
+		res[i] = &AllocationLogEntry{
+			Message:   log.Message,
+			Timestamp: log.Timestamp,
+			Count:     log.Count,
+		}
+		i++
+	}
+	sort.SliceStable(res, func(i, j int) bool {
+		return res[i].Timestamp.Before(res[j].Timestamp)
```

Comment:
Can we sort outside of the lock? This could even be postponed to the web handling when it is all done out of band.